xarray/core/dataset.py

```diff
@@ -2577,63 +2577,76 @@ def from_dataframe(cls, dataframe):
             obj[name] = (dims, data)
         return obj
 
-    def to_dask_dataframe(self, set_index=False):
+    def to_dask_dataframe(self, dim_order=None, set_index=False):
         """
         Convert this dataset into a dask.dataframe.DataFrame.
 
-        Both the coordinate and data variables in this dataset form
+        The dimensions, coordinates and data variables in this dataset form
         the columns of the DataFrame.
 
-        If set_index=True, the dask DataFrame is indexed by this dataset's
-        coordinate.  Since dask DataFrames to not support multi-indexes,
-        set_index only works if there is one coordinate dimension.
+        Arguments
+        ---------
+        dim_order : list, optional
+            Hierarchical dimension order for the resulting dataframe. All
+            arrays are transposed to this order and then written out as flat
+            vectors in contiguous order, so the last dimension in this list
+            will be contiguous in the resulting DataFrame. This has a major
+            influence on which operations are efficient on the resulting dask
+            dataframe.
+
+            If provided, must include all dimensions on this dataset. By
+            default, dimensions are sorted alphabetically.
+        set_index : bool, optional
+            If set_index=True, the dask DataFrame is indexed by this dataset's
+            coordinate. Since dask DataFrames to not support multi-indexes,
+            set_index only works if the dataset only contains one dimension.
+
+        Returns
+        -------
+        dask.dataframe.DataFrame
         """
 
         import dask.dataframe as dd
 
-        ordered_dims = self.dims
-        chunks = self.chunks
-
-        # order columns so that coordinates appear before data
-        columns = list(self.coords) + list(self.data_vars)
-
-        data = []
-        for k in columns:
-            v = self._variables[k]
-
-            # consider coordinate variables as well as data varibles
-            if isinstance(v, xr.IndexVariable):
-                v = v.to_base_variable()
+        if dim_order is None:
+            dim_order = list(self.dims)
+        elif set(dim_order) != set(self.dims):
+            raise ValueError(
+                'dim_order {} does not match the set of dimensions on this '
+                'Dataset: {}'.format(dim_order, list(self.dims)))
 
-            # ensure all variables span the same dimensions
-            v = v.set_dims(ordered_dims)
+        ordered_dims = OrderedDict((k, self.dims[k]) for k in dim_order)
 
-            # ensure all variables have the same chunking structure
-            if v.chunks != chunks:
-                v = v.chunk(chunks)
```

Comment:
@jmunroe was there was reason why you didn't just chunk everything here?

Reply:
On the assumption (probably mistaken) that there was a cost to calling .chunk(chunks) on a variable that already had that chunking structure.  If that assumption was not correct, then, yes, everything could just be chunked.

Reply:
Rechunk is nearly free if chunks are unchanged -- it actually returns the same dask array object.